R/rd-find-link-files.R

```diff
@@ -0,0 +1,90 @@
+
+#' Find the Rd file of a topic
+#'
+#' @param pkg Package to search in, or `NA` if no package was specified.
+#'   If the same as the dev package, then we treat it as `NA`.
+#' @param topic Topic to search for. This is the escaped, so it is `"\%\%"` and
+#'   not `"%%"`.
+#' @param tag The roxy tag object that contains the link. We use this for
+#'   better warnings, that include the file name and line number (of the tag).
+#' @return String. File name to link to.
+#'
+#' @details
+#' If `pkg` is `NA` or the package being documented, we'll just leave the
+#' topic alone.
+#'
+#' If `pkg` is not `NA` and not the package being documented (the _dev_
+#' package), then we need to be able to find the Rd file. If we can't, that's
+#' a warning and the link is left untouched. This typically happens when the
+#' linked package is not installed or cannot be loaded.
+#'
+#' @noRd
+
+find_topic_filename <- function(pkg, topic, tag = NULL) {
+  if (is.na(pkg) || identical(roxy_meta_get("current_package"), pkg)) {
+    topic
+  } else {
+    try_find_topic_in_package(pkg, topic, tag = tag)
+  }
+}
+
+#' Find a help topic in a package
+#'
+#' This is used by both `find_topic_filename()` and
+#' `format.rd_section_reexport()` that creates the re-exports page. The error
+#' messages are different for the two, so errors are not handled here.
+#'
+#' @param pkg Package name. This cannot be `NA`.
+#' @inheritParams find_topic_filename
+#' @return File name if the topic was found, `NA` if the package could be
+#'   searched, but the topic was not found. Errors if the package cannot be
+#'   searched. (Because it is not installed or cannot be loaded, etc.)
+#'
+#' @noRd
+
+find_topic_in_package <- function(pkg, topic) {
+  # This is needed because we have the escaped text here, and parse_Rd will
+  # un-escape it properly.
+  raw_topic <- str_trim(tools::parse_Rd(textConnection(topic))[[1]][1])
```

Comment:
Does this leave a dangling connection?

Reply:
I think read only text connections are gracefully closed, without a warning:
```r
> readLines(textConnection("foo\n"))
[1] "foo" ""   
> gc()
          used (Mb) gc trigger (Mb) max used (Mb)
Ncells  665032 35.6    1496105   80  1340143 71.6
Vcells 1122545  8.6    8388608   64  2094520 16.0
> gc()
          used (Mb) gc trigger (Mb) max used (Mb)
Ncells  665014 35.6    1496105   80  1340143 71.6
Vcells 1122532  8.6    8388608   64  2094520 16.0
> gc()
          used (Mb) gc trigger (Mb) max used (Mb)
Ncells  665014 35.6    1496105   80  1340143 71.6
Vcells 1122532  8.6    8388608   64  2094520 16.0
```

But I'll clean it up, just in case. It does not hurt.